libs/server/ArgSlice/ScratchAllocationManager.cs

```diff
@@ -0,0 +1,292 @@
+        /// Create an <see cref="ArgSlice"/> from the given ReadOnlySpan
+        /// </summary>
+        /// <param name="bytes">Input bytes</param>
+        /// <returns>Created <see cref="ArgSlice"/></returns>
+        public ArgSlice CreateArgSlice(ReadOnlySpan<byte> bytes)
+        {
+            ExpandScratchBufferIfNeeded(bytes.Length);
+
+            var retVal = new ArgSlice(currScratchBuffer.scratchBufferHead + currScratchBuffer.scratchBufferOffset, bytes.Length);
+            bytes.CopyTo(retVal.Span);
+
+            currScratchBuffer.scratchBufferOffset += bytes.Length;
+
+            return retVal;
+        }
+
+
+        /// <summary>
+        /// Create an <see cref="ArgSlice"/> in UTF8 format from the given string
+        /// </summary>
+        /// <param name="str">Input string</param>
+        /// <returns>Created <see cref="ArgSlice"/></returns>
+        public ArgSlice CreateArgSlice(string str)
+        {
+            var length = Encoding.UTF8.GetByteCount(str);
+            ExpandScratchBufferIfNeeded(length);
+
+            var retVal = new ArgSlice(currScratchBuffer.scratchBufferHead + currScratchBuffer.scratchBufferOffset, length);
+            Encoding.UTF8.GetBytes(str, retVal.Span);
+            currScratchBuffer.scratchBufferOffset += length;
+
+            return retVal;
+        }
+
+        /// <summary>
+        /// Create an <see cref="ArgSlice"/> of specified length, leaves contents as is
+        /// </summary>
+        /// <param name="length">Length of slice</param>
+        /// <returns>Created <see cref="ArgSlice"/></returns>
+        public ArgSlice CreateArgSlice(int length)
+        {
+            ExpandScratchBufferIfNeeded(length);
+
+            var retVal = new ArgSlice(currScratchBuffer.scratchBufferHead + currScratchBuffer.scratchBufferOffset, length);
+            currScratchBuffer.scratchBufferOffset += length;
+            Debug.Assert(currScratchBuffer.scratchBufferOffset <= currScratchBuffer.Length);
+            return retVal;
+        }
+
+        void SaveCurrentBufferAsBackupIfNeeded()
+        {
+            if (backupScratchBufferMaxSize != -1 && currScratchBuffer.Length <= backupScratchBufferMaxSize &&
+                (backupScratchBuffer.IsDefault || backupScratchBuffer.Length < currScratchBuffer.Length))
+            {
+                totalLength += backupScratchBuffer.IsDefault
+                    ? currScratchBuffer.Length
+                    : (backupScratchBuffer.Length - currScratchBuffer.Length);
+
+                backupScratchBuffer = currScratchBuffer;
+            }
+        }
+
+        void ExpandScratchBufferIfNeeded(int requiredLength)
+        {
+            if (currScratchBuffer.IsDefault || requiredLength > currScratchBuffer.Length - currScratchBuffer.scratchBufferOffset)
+                ExpandScratchBuffer(requiredLength);
+        }
+
+        void ExpandScratchBuffer(int requiredLength)
+        {
+            ScratchBuffer newScratchBuffer;
```

Comment:
since the buffer is meant to be used single threaded, there is no need to waste stack space on a newScratchBuffer.